Mathlib/Analysis/ODE/PicardLindelof.lean

```diff
@@ -10,397 +10,716 @@ import Mathlib.Topology.MetricSpace.Contracting
-that the fixed point of this map is the solution of the corresponding ODE.
-
-More precisely, some iteration of this map is a contracting map. -/
-def next (f : FunSpace v) : FunSpace v where
-  toFun t := v.x₀ + ∫ τ : ℝ in v.t₀..t, f.vComp τ
-  map_t₀' := by simp only [integral_same, add_zero]
-  lipschitz' := LipschitzWith.of_dist_le_mul fun t₁ t₂ => by
-    rw [dist_add_left, dist_eq_norm,
-      integral_interval_sub_left (f.intervalIntegrable_vComp _ _) (f.intervalIntegrable_vComp _ _)]
-    exact norm_integral_le_of_norm_le_const fun t _ => f.norm_vComp_le _
-
-theorem next_apply (t : Icc v.tMin v.tMax) : f.next t = v.x₀ + ∫ τ : ℝ in v.t₀..t, f.vComp τ :=
-  rfl
-
-theorem dist_next_apply_le_of_le {f₁ f₂ : FunSpace v} {n : ℕ} {d : ℝ}
-    (h : ∀ t, dist (f₁ t) (f₂ t) ≤ (v.L * |t.1 - v.t₀|) ^ n / n ! * d) (t : Icc v.tMin v.tMax) :
-    dist (next f₁ t) (next f₂ t) ≤ (v.L * |t.1 - v.t₀|) ^ (n + 1) / (n + 1)! * d := by
-  simp only [dist_eq_norm, next_apply, add_sub_add_left_eq_sub, ←
-    intervalIntegral.integral_sub (intervalIntegrable_vComp _ _ _)
-      (intervalIntegrable_vComp _ _ _),
-    norm_integral_eq_norm_integral_uIoc] at *
+    ‖α t - x₀‖ ≤ ‖α t - α t₀‖ + ‖α t₀ - x₀‖ := norm_sub_le_norm_sub_add_norm_sub ..
+    _ ≤ L * |t.1 - t₀.1| + r := by
+      apply add_le_add _ <| mem_closedBall_iff_norm.mp α.mem_closedBall₀
+      rw [← dist_eq_norm]
+      exact α.lipschitzWith.dist_le_mul t t₀
+    _ ≤ L * max (tmax - t₀) (t₀ - tmin) + r := by
+      gcongr
+      exact abs_sub_le_max_sub t.2.1 t.2.2 _
+    _ ≤ a - r + r := add_le_add_right h _
+    _ = a := sub_add_cancel _ _
+
+lemma compProj_mem_closedBall
+    (α : FunSpace t₀ x₀ r L) (h : L * max (tmax - t₀) (t₀ - tmin) ≤ a - r) {t : ℝ} :
+    α.compProj t ∈ closedBall x₀ a := by
+  rw [compProj_apply]
+  exact α.mem_closedBall h
+
+end
+
+/-! ## Contracting map on the space of Lipschitz functions -/
+
+section
+
+variable [NormedSpace ℝ E]
+  {f : ℝ → E → E} {tmin tmax : ℝ} {t₀ : Icc tmin tmax} {x₀ x y : E} {a r L K : ℝ≥0}
+
+/-- The integrand in `next` is continuous. -/
+lemma continuousOn_comp_compProj (hf : IsPicardLindelof f t₀ x₀ a r L K) (α : FunSpace t₀ x₀ r L) :
+    ContinuousOn (fun t' ↦ f t' (α.compProj t')) (Icc tmin tmax) :=
+  continuousOn_comp
+    (continuousOn_prod_of_continuousOn_lipschitzOnWith' (uncurry f) K hf.lipschitzOnWith
+      hf.continuousOn)
+    α.continuous_compProj.continuousOn
+    fun _ _ ↦ α.mem_closedBall hf.mul_max_le
+
+/-- The integrand in `next` is integrable. -/
+lemma intervalIntegrable_comp_compProj (hf : IsPicardLindelof f t₀ x₀ a r L K)
+    (α : FunSpace t₀ x₀ r L) (t : Icc tmin tmax) :
+    IntervalIntegrable (fun t' ↦ f t' (α.compProj t')) volume t₀ t := by
+  apply ContinuousOn.intervalIntegrable
+  apply α.continuousOn_comp_compProj hf |>.mono
+  exact uIcc_subset_Icc t₀.2 t.2
+
+/-- The map on `FunSpace` defined by `picard`, some `n`-th interate of which will be a contracting
+map -/
```

Comment:
(Just because you end some of your docstrings with a dot.)
```suggestion
/-- The map on `FunSpace` defined by `picard`, some `n`-th interate of which will be a contracting
map. -/
```